File: Itido-BMS/dart/bms_dart/lib/src/models/work.dart
import 'package:flutter/material.dart';
import 'package:bms_dart/models.dart';
import 'package:dart_packages/date_time_operations.dart' as dtOps;

import 'date_time_converter.dart';

class Work {
  int id;
  String note;
  DateTime date;
  int startTimeMins;
  int endTimeMins;
  int breakMins;
  bool isVisible;
  bool isInvited;
  int inviteCount;
  bool canRegisterWork;

  Contract contract;
  Project project;
  WorkReplacement workReplacement;
  WorkRegistration workRegistration;

  int workContractId;

  // int _projectId;
  // int _contractId;
  // int workTypeId;
  //int workReplacementId;
  //int workRegistrationId;

  //User user;

  Work({
    // this.workContractId,
    // this.contractId,
    // this._projectId,
    // this.workTypeId,
    this.note,
    this.date,
    this.startTimeMins,
    this.endTimeMins,
    this.breakMins,
    //this.workReplacementId,
    this.isVisible,
    // int projectId,
    // this.workRegistrationId,
  });

  Work.fromJson(json)
      : this.id = json['id'],
        this.workContractId = json['workContractID'],
        // this._contractId = json['contractId'],
        // this._projectId = json['projectID'],
        // this.workTypeId = json['workTypeId'],
        this.note = json['note'],
        this.date = toDateTime(json['date']),
        this.startTimeMins = json['startTimeMins'],
        this.endTimeMins = json['endTimeMins'],
        this.breakMins = json['breakMins'],
        //this.workReplacementId = json['workReplacementId'],
        this.isVisible = json['isVisible'],
        this.isInvited = json['isInvited'],
        this.inviteCount = json['inviteCount'],
        this.canRegisterWork = json['canRegisterWork'],
        // this.workRegistrationId = json['workRegistrationId'],

        this.project = Project.fromJson(json['projectItem_project']),
        this.contract = Contract.fromJson(
          json['contract'],
          rootJson: json,
          path: 'contract',
        ),
        this.workReplacement = WorkReplacement.fromJson(
          json['workReplacement'],
          rootJson: json,
          path: 'workReplacement',
        ),
        this.workRegistration = WorkRegistration.fromJson(
          json['workRegistration'],
          rootJson: json,
          path: 'workRegistration',
        );

  Map<String, dynamic> toMap() => {
        'id': this.id,
        'note': this.note,
        'date': this.date.toString(),
        'startTimeMins': this.startTimeMins,
        'endTimeMins': this.endTimeMins,
        'breakMins': this.breakMins,
        'isVisible': this.isVisible,
        'isInvited': this.isInvited,
        'inviteCount': this.inviteCount,
        // 'workContractId': this.workContractId,
        // 'projectId': this._projectId,
        // 'contractId': this._contractId,
        'modifications': 0,
      };

  bool get isRegistered => workRegistration != null;

  bool get isOwned => contract != null;

  bool get isOwnerAbsent => workReplacement != null;
  bool get isReplaced => workReplacement?.contract != null;

  bool get isTaken =>
      contract != null && workReplacement == null ||
      workReplacement != null && workReplacement.contract != null;

  bool get isUserAbsent =>
      workReplacement != null && workReplacement.contract == null;

  bool get isElegibleForRegistration =>
      !isRegistered && dtOps.compareDate(DateTime.now(), date) >= 0;

  bool get isLate =>
      !isRegistered && dtOps.compareDate(DateTime.now(), date) == 1;

  bool get isPartOfWorkContract => workContractId != null;

  User get user => isOwned && contract.user != null ? contract.user : null;
}

File: Itido-BMS/RentAPI/Rent/Models/TimePlanning/Work.cs
using System;
using System.Linq;
using System.Linq.Expressions;
using Rent.Models.Projects;
using System.Collections.Generic;
using Rent.Helpers;

namespace Rent.Models.TimePlanning
{
    public class Work
    {
        public int ID { get; set; }
        public int? ContractID { get; set; }
        public int? WorkContractID { get; set; }

        //public int? LocationID { get; set; }

        public string Note { get; set; }
        public DateTime Date { get; set; }

        public int Modifications { get; set; }

        public short StartTimeMins { get; set; }
        public short EndTimeMins { get; set; }
        public short BreakMins { get; set; }
        //public int? WorkReplacementID { get; set; }
        public bool IsVisible { get; set; }
        // public int? WorkRegistrationID { get; set; }
        public virtual Contract Contract { get; set; }
        //public virtual Location Location { get; set; }
        public virtual WorkReplacement WorkReplacement { get; set; }
        public virtual WorkRegistration WorkRegistration { get; set; }

        public virtual ICollection<WorkInvitation> WorkInvitations { get; set; }

        public int? ProjectItemID { get; set; }
        public virtual ProjectItem ProjectItem { get; set; }

        public static Expression<Func<Work, dynamic>> BasicDTO()
        {
            return v => v != null ?
            new
            {
                v.ID,
                v.Note,
                v.Date,
                v.StartTimeMins,
                v.EndTimeMins,
                v.BreakMins,
                v.IsVisible,
            } : null;
        }

        public static Expression<Func<Work, dynamic>> StandardDTO(int requester, string requesterRole)
        {
            return v => v != null ?
            new
            {
                v.ID,
                v.Note,
                v.Date,
                v.StartTimeMins,
                v.EndTimeMins,
                v.BreakMins,
                v.IsVisible,

                // project 
                ProjectItem = ProjectItem.BasicDTO().Compile()(v.ProjectItem),
                ProjectItem_project = v.ProjectItem != null ? Project.BasicDTO().Compile()(v.ProjectItem.Project) : null,

                // contract
                Contract = Contract.BasicDTO().Compile()(v.Contract),
                Contract_user = v.Contract != null ? User.BasicDTO().Compile()(v.Contract.User) : null,

                // workReplacement
                WorkReplacement =
                    WorkReplacement.BasicDTO().Compile()(v.WorkReplacement),
                WorkReplacement_contract =
                    v.WorkReplacement != null ? Contract.BasicDTO().Compile()(v.WorkReplacement.Contract) : null,
                WorkReplacement_contract_user =
                    v.WorkReplacement != null && v.WorkReplacement.Contract != null ? User.BasicDTO().Compile()(v.WorkReplacement.Contract.User) : null,
                WorkReplacement_absence = Absence.BasicDTO(requester, requesterRole).Compile()(v.WorkReplacement.Absence),
                WorkReplacement_absence_absenceReason = AbsenceReason.BasicDTO().Compile()(v.WorkReplacement.Absence.AbsenceReason),

                // workRegistration
                WorkRegistration = WorkRegistration.BasicDTO().Compile()(v.WorkRegistration),
                WorkRegistration_request = Request.BasicDTO(requester, requesterRole).Compile()(v.WorkRegistration.Request, v.WorkRegistration.Request.Creator),

                IsInvited = v.WorkInvitations.Any(inv => inv.Contract.UserID == requester),
                InviteCount = v.WorkInvitations.Count,

                CanRegisterWork = v.WorkRegistration == null && (requesterRole == "Admin" || (v.Contract != null && (v.WorkReplacement != null && v.WorkReplacement.Contract != null && v.WorkReplacement.Contract.UserID == requester) || (v.WorkReplacement == null && v.Contract.UserID == requester))),
            } : null;
        }

        public static Expression<Func<Work, dynamic>> DetailedDTO(int requester, string requesterRole)
        {
            return v => v != null ?
            new
            {
                v.ID,
                v.Note,
                v.Date,
                v.StartTimeMins,
                v.EndTimeMins,
                v.BreakMins,
                v.IsVisible,

                // project 

                ProjectItem = ProjectItem.BasicDTO().Compile()(v.ProjectItem),
                ProjectItem_project = v.ProjectItem != null ? Project.BasicDTO().Compile()(v.ProjectItem.Project) : null,

                // contract
                Contract = Contract.BasicDTO().Compile()(v.Contract),
                Contract_user = v.Contract != null ? User.BasicDTO().Compile()(v.Contract.User) : null,

                // workReplacement
                WorkReplacement =
                    WorkReplacement.BasicDTO().Compile()(v.WorkReplacement),
                WorkReplacement_contract =
                    v.WorkReplacement != null ? Contract.BasicDTO().Compile()(v.WorkReplacement.Contract) : null,
                WorkReplacement_contract_user =
                    v.WorkReplacement != null && v.WorkReplacement.Contract != null ? User.BasicDTO().Compile()(v.WorkReplacement.Contract.User) : null,
                WorkReplacement_absence = v.WorkReplacement != null ? Absence.BasicDTO(requester, requesterRole).Compile()(v.WorkReplacement.Absence) : null,
                WorkReplacement_absence_absenceReason = AbsenceReason.BasicDTO().Compile()(v.WorkReplacement.Absence.AbsenceReason),

                // workRegistration
                WorkRegistration = WorkRegistration.BasicDTO().Compile()(v.WorkRegistration),
                WorkRegistration_request = v.WorkRegistration != null ? Request.BasicDTO(requester, requesterRole).Compile()(v.WorkRegistration.Request, v.WorkRegistration.Request.Creator) : null,

                IsInvited = v.WorkInvitations.Any(inv => inv.Contract.UserID == requester),
                InviteCount = v.WorkInvitations.Count,

                CanRegisterWork = v.WorkRegistration == null && (requesterRole == "Admin" || (v.Contract != null && (v.WorkReplacement != null && v.WorkReplacement.Contract != null && v.WorkReplacement.Contract.UserID == requester) || (v.WorkReplacement == null && v.Contract.UserID == requester))),
            } : null;
        }
    }
}